mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib
+
+import breeze.linalg.{ Vector => BV }
+
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.rdd.RDD
+
+package object base {
```

Comment:
@mengxr

I tried to build the Spark master branch, but was unsuccessful.

So, instead of integrating my Bregman clusterer into the latest master
branch of Spark, I integrated them into my public GitHub project
<https://github.com/derrickburns/generalized-kmeans-clustering>.

Per your suggestion, this   version supports sparse vectors.

Included is support for several Bregman divergences including: squared
Euclidean distance, Kullback Leibler divergence, logistic loss, and
generalized I-divergence.

I created a general BregmanDivergence trait and a separate BregmanPointOps
trait.  When mixed in together, one gets a specific BregmanPointOps object
that computes Bregman divergences very efficiently, caching one value per
point and two values per center (per your observation).  This version is
easier to understand that the previous one.

Included also are three different implementations of K-Means clustering: 1)
one that caches distances (CachingKMeans) 2) one that tracks which clusters
moved and which points are assigned to which clusters and the distance to
the closest cluster (TrackingKMeans); and, 3) one that recomputes all
clusters and distances (SingleKMeans).

I have found the the TrackingKMeans clusterer has the best performance on
last data sets.

I have also found that maintaining cluster centers in homogeneous
coordinates is more efficient than using inhomogeneous coordinates for
moderate to high dimensional data. There is no penalty for low dimensional
data.

If/when you include such changes in Spark, I would appreciate being
included in the list of contributors.

Regards,

Derrick Burns



On Sun, Jan 11, 2015 at 10:52 AM, Derrick Burns <derrickrburns@gmail.com>
wrote:

> I'm willing to take one last crack at contributing the Bregman
> modifications to Spark. Here is an layering of the pull requests:
>
> 1) modifications to BLAS to support arbitrary functions and axpy on
> SparseVectors
> 2) definition of PointOps trait and various implementations for different
> divergences
> 3) rewrite of clusterer to use PointOps trait but only using the squared
> Euclidean distance implementation
> 4) generalization of clustering interface to select different divergences.
> 5) store seeds
> 6) maintain exactly K clusters by splitting clusters heuristicly.
>
> I can submit PRs for 1-4.
>
> Thoughts?
>
> Sent from my iPhone
>
> On Jan 8, 2015, at 4:00 PM, Xiangrui Meng <notifications@github.com>
> wrote:
>
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/package.scala
> <https://github.com/apache/spark/pull/2634#discussion-diff-22693266>:
>
> > + *
> > + * Unless required by applicable law or agreed to in writing, software
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib
> > +
> > +import breeze.linalg.{ Vector => BV }
> > +
> > +import org.apache.spark.mllib.linalg.Vector
> > +import org.apache.spark.rdd.RDD
> > +
> > +package object base {
>
> Though private, having some doc would make the code easy to understand.
> For example,
>
>    1. What does FP mean?
>    2. inh?
>    3. ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634/files#r22693266>.
>
>